vitis/hlsBs/man1/hlsRun.1
.TH hlsRun 1 "May 2026" "v1.0" "HLS Command Manual"
.SH NAME
hlsRun \- Runs the \fBcsim\fR, \fBsynthesis\fR, \fBcosim\fR, \fBpackage\fR and
\fBip\fR on one or more of the components in the workspace

.SH SYNOPSIS
.B hlsRun [COMPONENTS] [OPTIONS]

.SH DESCRIPTION
.TP
.B hlsRun
This command, together with its more specialized and simpler variations,
\fBhlsExe\fR and \fBhlsGdb\fR, will be the most often used commands. Other
\fBhls\fR commands (\fBhlsWs\fR, \fBhlsCfg\fR and \fBhlsComp\fR) are seldom
used after creating and managing the configurations and their corresponding
components.


.SH COMPONENTS
A list of the names of the components, either fully specified or with
wildcards.

Similarly, while all the various stages, \fBcsim\fR, \fBsynthesis\fR, \fBcosim\fR,
\fBpackage\fR and \fBip\fR can be run simultaneously, except for regeneration
of a known working project, due to the time and amount of generated output,
it is more practical to run them separately.

.SH OPTIONS
.TP
\fB\-h, \-\-help\fR
Displays this message\

.TP
\fB\-\-verbose
Enables more extensive output which is useful for diagnostics.  The default
is 'False'.

.TP
\fB\-w, \-\-workspace = \fIWORKSPACE\fR
The target workspace. Although typically specified in the project file, this
overrides that defintion and is useful for one-off/throw-away experimentation.
The environment variable \fBHLSBS_WORKSPACE\fR can also be used.

.TP
\fB\-\-ip_root = \fIP_ROOT\fR
The ip root directory. Although typically specified in the project file, this
overrides that defintion and is useful for one-of/throw-away experimentation.
The environment variable \fBHLSBS_IP_ROOT\fR can also be used.


.TP
\fB\-\-components=...\fR
A list of component names.  Functionally this acts the same as the positional
\fBCOMPONENTS\fR arguments, but due to shell's lack of a formal binding syntax
for lists and the shell's handling of wildcards for positional arguments this
is more explicit and reliable. \fB\-\-components\fR and positional
parameters can be used separately or together.

.TP
\fB\-\-project = \fIPROJECT_FILE\fR
A project file is the standard method used by all the \fBhls\fR build utilities
of capturing the stable command line options in a file that can be saved in GIT.
For existing projects, the formality of a project file is not necessary. There
are simpler ways. Only for more complex projects is a project file needed. It is
mainly for consistency with other \fBhls\fR build utilities. Default is the
environment variable \fIHLSBS_PROJECT\fR.

.TP
\fB\-\-no-project
This prevents the use of a project file which may or may not exist.
Typically used when one wishes to specify all the relevant parameters (in this
utility these are \fBworkspace\fR and \fBconfigurations\fR and, possibly
\fBcomponents\fR if one wishes something other than the default. A common usage is
in one-off experimentation.

.SH ACTIONS
.TP
\fB\-\-list\fR
Lists the viable components.  This is provided as a convenience when the
exact names of components are inevitably forgotten.

.TP
\fB\-\-csim\fR = (make, clean, run)
Executes any combination of  making, cleaning, or running the csim.exe that
makes sense. See the \fB--fast\fR and \fB--slow\fR options which are relevant
to this option.

.TP
\fB\-\-synthesis\fR
Runs the synthesis stage

.TP
\fB\-\-cosim\fR = (clean, make, run)
Executes any combination of  making, cleaning, running  the cosim stage that
makes sense. There is no  \fB--fast\fR option for the cosim stage.

.TP
\fB\-\-package\fR
Executes the packaging stage.

.TP
\fB\-\-implementation\fR
Executes the implementation stage.

.TP
\fB\-\-ip\fR
Run the ip stage. If not using a project file to capture the \fBip\fR options,
see the following list of \fBip\fR options.

\fB\-\-ip-file\fR=\fIFILE_PATH_NAME\fR

An explicit or template to specify where the output of the \fB--ip\fR
is written.  The symbolic's \fBvitis_version\fR and \fBcmp_name\fR can
be used.

This may be a full file path specification or just the
directory. A missing file name and file extension will be first be
taken from \fBip-name\fR.  If still missing, the file name defaults to
\fB{cmp_name}\fR and the extension to \fB.zip\fR.

\fB\-\-ip-name\fR=\fINAME\fR

The file name holding the IP products. This can be used to provide or replace
the file name and extension in \fBip-file\fR.

\fB\-\-ip-family\fR=\fIFAMILY\fR

The list of Xilinx FPGAs families that are compatible with this product.


.TP
\fB\-\-dcp\fR
Run the dcp renaming stage. If not using a project file to capture the
\fBip\fR options, see the following list of \fBdcp\fR options.

\fB\-\-dcp-file\fR=\fIFILE_PATH_NAME\fR

An explicit or template to specify where the output of the \fB--ip\fR
is written.  The symbolic's \fBvitis_version\fR and \fBcmp_name\fR can
be used.

This may be a full file path specification or just the
directory. Any missing file name and file extension will be first be
taken from \fBip-name\fR.  If still missing, the file name defaults to
\fB{cmp_name}\fR and the extension to \fB.dcp\fR.

\fB\-\-dcp-name\fR=\fINAME\fR

The file name holding the DCP products. This can be used to provide or replace
the file name and extension in \fBdcp-file\fR.

.TP
\fB\-\-all\fR = (clean, make, run)
Executes all stages except the packaging stage since the
\fBpackaging\fR stage is subset of the \fBimplementation\fR stage. It
can be explicitly included with the \fB\-\-package option\fR.  The
option list, i.e. \fBclean\fR, \fBmake\fR, \fBrun\fR, selects these
options for both the \fBcsim\fR and \fBcosim\fR.  An option's list
provided with an explicit \fB\-\-csim\fR or \fB\-\-cosim\fR takes
precedence.

\fB\-\-exclude\fR=(csim, synthesis, cosim, package, implementation, ip)
When used with \fB\-\-all\fR, removes that stage from the \fBall\fR list.

.TP
\fB\-\-dry-run\fR
Only reports the actions the command would do, but does not run the command.


.SH Project File
A project file is used to capture values of various options that could be
specified on the command line. This saves both typing and remembering what they
are. Given that \fBhlsComp\fR is generally used in very simple existing projects,
no project file or a very minimal one is needed specifying
    1. The workspace
    2. The configuration file(s)
    3. The mapping of a configuration file name to a component


.SH Component Selection
Selected components are ones that exist in the workspace and match the component
parameters. This is different from the previous commands which categorized the
components as
\fBexisting\fR, \fBmissing\fR or \fBcruft\fR. Since only existing components can
be targeted, the only ambiguity is those in the \fBcruft\fR category, i.e.
rogue components, which by entropy,  have somehow snuck their way into the
workspace. To keep this command simple, the burden falls on the user to use
either \fBhlsCfg\fR or \fBhlsComp\fR to eliminate them or the selection criteria
to avoid them.

.SH Overriding with Command Line Options
The \fBhls\fR utilities are constructed so parameters can be specified at 3
levels depending on their permanency
    1. In the project file
    2. Directly overridden on the command line
    3. With parameter files on the command line

The \fBproject file\fR is the most permanent and \fBoverriding\fR directly on
the command line the least. The disadvantage of overriding directly on the
command line is that, except for command recall, they are lost to history and even
then, they may be seen as just a random collection of options.

The third option offers something in between, allowing one to capture
command line options in a file.  This allows the options some degree of
permanency and, since it is a file, it can be given a meaningful name. The
disadvantage is, due to the complexity in some utilities, not all options offered
in the project file are available.

The command line syntax for a parameter file is \fI@<MY_OPTIONS>\fR, i.e. the
file name preceded by an \fB@\fR. The file is trivially formatted, one option
(no spaces) per line. These act just as if the contents had been placed in
the command line where the parameter file is.

See the example below.

.SH Examples:

  These examples specify the target components using the quoted positional
  arguments, but the recommendation is to use the \fI--components=\fR (or
  the abbreviation \fI--comp=\fR. It is less error-prone.

  All these can include a \fI--dry-run\fR option added to see what would have
  been done.


  Example 1
  $ hlsRun '*' --csim=m  ..or..
  $ hlsRun --csim=m

  Runs the make csim on all known components.

  Example 2
  $ hlsRun 'avg*' --csim   ..or..
  $ hlsRun 'avg*' --csim=clean,make,run ..or..
  $ hlsRun 'avg*  --csim=c,m,r

  These are all equivalent, running a clean, make and run on any components
  satisfying the wildcard 'avg*'

  Example 3
  $ hlsRun --components=avg*,dense --csim=r --syn --cosim  ..or..
  $ hlsRun @AvgDense.txt --csim=r --syn --cosim

  Runs the run, synthesis and cosim stages on all components satisfying the
  wildcard 'avg*' and the explicit 'dense'.

  Assuming \fbAvgDense.txt\fR contains
  --components=avg*,dense

  Example 4
  $ hlsRun 'avg*' --ip  --ip-file=/tmp/me/{vitis-version}/{cmp_name}

  Runs the ip stage placing the output in the specified file. The ip-file
  overrides any definition in the project file.  Handy for testing.

  Example 5
  $ hlsRun '*' --all

  Runs all the stages on all the components.  This is recommended only when
  doing batch processing. For interactive usage
     1.  The amount of output can be overwhelming
     2.  It could take a long time

.SH AUTHOR
Written by JJRussell


vitis/hlsBs/vitispy/hlsRun.py
# ----------------------------------------------------------------------------
# This file is part of the 'SLAC Firmware Standard Library'. It is subject to
# the license terms in the LICENSE.txt file found in the top-level directory
# of this distribution and at:
#    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
# No part of the 'SLAC Firmware Standard Library', including this file, may be
# copied, modified, propagated, or distributed except according to the terms
# contained in the LICENSE.txt file.
# ----------------------------------------------------------------------------

import argparse
import sys
import os
import runpy

# -------------------------------------------------------
# Add the VITIS HLS python paths, remove nonexistent ones
# ------------------------------------------
runpy.run_path(os.getenv('HLSBS_IMPORT_PYPATHS'),
               run_name='add_paths ' + str(sys.path))

import vitispy.files as files
from vitispy.project import Project
from vitispy.printer import Printer
from vitispy.workspace import Workspace
from vitispy.dry_run import DryRun
from vitispy.run import Run
from vitispy.manpage import display_manpage

# ------------------------------------------------------------------------------
class SplitArgs(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        if values:
            setattr(namespace, self.dest, values[0].split(','))
        else:
            setattr(namespace, self.dest, [])
        return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def add_arguments(parser):

    parser.add_argument('-h',
                        '--help',
                        action='store_true',
                        help='Show custom help')

    Workspace.add_arguments(parser)

    parser.add_argument('parameters',
                        help="The list of components",
                        nargs='*',
                        default=None)

    parser.add_argument('--verbose',
                        help='Minimum output',
                        action='store_true',
                        default=False)

    parser.add_argument('--components',
                        nargs='*',
                        default=None,
                        help='The target component(s)')

    parser.add_argument('--list',
                        help='List available components to run',
                        action='store_true',
                        default=False)

    parser.add_argument('--csim',
                        help='run csim',
                        nargs='*',
                        const='--no_value--',
                        action=SplitArgs)

    parser.add_argument('--synthesis',
                        help='run synthesis',
                        action='store_true',
                        default=False)

    parser.add_argument('--cosim',
                        help='run cosim',
                        nargs='*',
                        const='--no_value--',
                        action=SplitArgs)

    parser.add_argument('--package',
                        help='run package',
                        action='store_true',
                        default=False)

    parser.add_argument('--implementation',
                        help='run implementation',
                        action='store_true',
                        default=False)

    parser.add_argument('--ip',
                        help='create ip packaging',
                        nargs='*',
                        const='--no_value--',
                        action=SplitArgs)

    parser.add_argument('--all',
                        help='run all',
                        nargs='*',
                        const='--no_value--',
                        action=SplitArgs)

    parser.add_argument('--exclude',
                        help='list of run options to exclude')

    parser.add_argument('--slow',
                        help='make using the fast method',
                        action='store_true',
                        default=False)

    DryRun.add_arguments(parser)

    Project.    add_arguments(parser)
    Project.Ip. add_arguments(parser)

    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def report(verbose, printer, project, workspace, components, run):

    printer.separator('=')
    if run == 'List':
        run = None
        printer.line('List by Component', '')

    project.git = Project.Git(project.root)

    printer.line("Workspace", workspace)
    printer.line('Component', components)
    project.print_project(printer, verbose)

    print()
    if run:
        run.print(printer)
    printer.separator('-', 40)

    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def merge(x, y):
    '''
   Merges x, y into a list, where x,y can either be a list or a comma separated
   string
    '''
    # --------------------------------------------------------------------------
    if x:
        # Ensure X is a list
        if len(x) == 1:
            lst = x[0].split(',')
        else:
            lst = x
    else:
        lst = []

    if y:
        if len(y) == 1:
            lst += y[0].split(',')
        else:
            lst += y

    return lst
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
class Options:
    def __init__(self,
                 all,
                 exclude,
                 csim,
                 synthesis,
                 cosim,
                 package,
                 implementation,
                 ip,
                 slow,
                 dry_run,
                 verbose):

        self.slow = slow
        self.dry_run = dry_run
        self.verbose = verbose
        self.stages = 0

        if all is not None:
            self.stages = Run.Stage.All

        else:
            if csim is not None:
                self.stages |= Run.Stage.CSim
            if synthesis:
                self.stages |= Run.Stage.Synthesis
            if cosim is not None:
                self.stages |= Run.Stage.CoSim
            if package:
                self.stages |= Run.Stage.Package
            if implementation:
                self.stages |= Run.Stage.Implementation

        if all is not None:
            if csim is None:
                csim = 'all'
            if cosim is None:
                cosim = 'all'
            if ip is None:
                ip = 'all'
            self. csim_set(csim)
            self.cosim_set(cosim)
            self.stages |= self.ip_set(ip)

        else:
            if self.stages & Run.Stage.CSim:
                self.csim_set(csim)

            if self.stages & Run.Stage.CoSim:
                self.cosim_set(cosim)

            self.stages |= self.ip_set(ip)

        # Exclude unwanted actions
        self.exclude = exclude
        if exclude:
            exclude_list = exclude.split(',')
            for item in exclude_list:
                unique = item[0:2]
                if (unique == 'cs'):
                    self.stages &= ~Run.Stage.CSim
                if (unique == 'sy'):
                    self.stages &= ~Run.Stage.Synthesis
                if (unique == 'co'):
                    self.stages &= ~Run.Stage.CoSim
                if (unique == 'pa'):
                    self.stages &= ~Run.Stage.Package
                if (unique == 'im'):
                    self.stages &= ~Run.Stage.Implementation
                if (unique == 'ip'):
                    self.stages &= ~Run.Stage.Ip_Zip
                if (unique == 'dc'):
                    self.stages &= ~Run.Stage.Ip_Dcp

        return
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    def csim_set(self, list):

        # Hold on to the original request
        msk = self.action_msk(list, Run.Msk.Make | Run.Msk.Run)
        self.csim = msk != 0
        self.csim_msk = msk

        # -------------------------------------------------------
        # Modify the mask for limitations of the slow HLS actions
        # The fast modification only disallows Clean | Run combo
        # -------------------------------------------------------
        self.csim_fast_msk = Run.Msk.CSimFastValid[msk]
        self.csim_slow_msk = Run.Msk.CSimSlowValid[msk]

        return
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    def cosim_set(self, list):

        # Hold on to the original request
        msk = self.action_msk(list, Run.Msk.Make | Run.Msk.Run)
        self.cosim = msk != 0

        # --------------------------------------------------
        # Modify the mask for limitations of the HLS actions
        # --------------------------------------------------
        self.cosim_msk = Run.Msk.CoSimValid[msk]

        return
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    @staticmethod
    def ip_set(ip_list):

        msk = 0

        if ip_list is None:
            return msk

        if not isinstance(ip_list, list):
            if ip_list == '--no_value--':
                msk = Run.Stage.Ip_Zip | Run.Stage.Ip_Dcp
                return msk
            else:
                ip_list = [ip_list]
        elif len(ip_list) == 0:
            msk = Run.Stage.Ip_Zip | Run.Stage.Ip_Dcp
            return msk

        if Options.is_in('z', ip_list):
            msk |= Run.Stage.Ip_Zip
        if Options.is_in('d', ip_list):
            msk |= Run.Stage.Ip_Dcp
        if Options.is_in('a', ip_list):
            msk = Run.Stage.Ip_Zip | Run.Stage.Ip_Dcp

        return msk
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    @staticmethod
    def is_in(s, list):
        n = len(s)
        for member in list:
            if s == member[0:n]:
                return True
        return False
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    @staticmethod
    def action_msk(list, default_msk):

        # If List is None, no action
        if list is None:
            return 0

        msk = 0

        # If empty list, return the default action mask
        if len(list) == 0:
            msk = default_msk
            return msk

        all = Options.is_in('a', list)
        if all:
            msk = Run.Msk.Clean | Run.Msk.Make | Run.Msk.Run
        else:
            if Options.is_in('c', list):
                msk |= Run.Msk.Clean
            if Options.is_in('m', list):
                msk |= Run.Msk.Make
            if Options.is_in('r', list):
                msk |= Run.Msk.Run

            # Default the action if clean or run not specified
            if not (msk & Run.Msk.Clean) and not (msk & Run.Msk.Run):
                msk |= Run.Msk.Make

        return msk
    # --------------------------------------------------------------------------
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def noComponentsFound(printer, workspace, components):

    import difflib
    printer.line(f"Warning: No components found using {components},"
                 "choices are", sep='*')

    # ----------
    # Candidates
    # ----------
    cmps = files.get_components(workspace, '*')
    cmp_names = []
    for cmp_path in cmps:
        cmp_name = os.path.splitext(os.path.split(cmp_path)[1])[0]
        cmp_names.append(cmp_name)

    # -----------------
    # Potential Matches
    # ----------------
    matches = []
    for cmp_name in components:
        ms = difflib.get_close_matches(cmp_name,
                                       cmp_names,
                                       n=4,
                                       cutoff=0.4)
        if len(ms):
            matches.extend(ms)

    icmp = 1
    cmps.sort()
    for cmp_name in cmp_names:
        sep = '<-----' if cmp_name in matches else ''
        printer.item(icmp, cmp_name, sep, '')
        icmp += 1

    if len(matches):
        print()
        matches.sort()
        printer.line("Perhaps you meant", '')
        icmp = 1
        for cmp_name in matches:
            printer.item(icmp, cmp_name, None)
            icmp += 1

    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def doit():
    parser = argparse.ArgumentParser(add_help=False,
                                     fromfile_prefix_chars='@')

    add_arguments(parser)
    args = parser.parse_args()
    if args.help:
        display_manpage(__file__)
        exit(0)

    components = merge(args.parameters, args.components)
    printer = Printer(18, 78, 20)
    needs = Project.Need.Workspace
    if args.ip is not None:
        needs |= (Project.Need.Products_Root
              |   Project.Need.Ip_Root)

    project = Project(needs,
                      args.project,
                      args.root,
                      args.products_root,
                      None,                # build_root not needed
                      args.workspace,
                      None,                # cfg_root   not needed
                      args.ip_dir)

    workspace = project.workspace
    if not workspace:
        print("ERROR: No workspace specified")
        return -1

    options = Options(args.all,
                      args.exclude,
                      args.csim,
                      args.synthesis,
                      args.cosim,
                      args.package,
                      args.implementation,
                      args.ip,
                      args.slow,
                      args.dry_run,
                      args.verbose)

    # ----------------------------
    # Setup IP family augmentation
    # ----------------------------
    ip = None
    if options.stages & (Run.Stage.Ip_Zip | Run.Stage.Ip_Dcp):
        if project.has_project_file:
            ip = project.get_ip()
            if ip:
                ip.replace(args)

        if not ip:
            ip = Project.Ip(dir=args.ip_dir,
                            zip_file=args.ip_zip_file,
                            dcp_file=args.ip_dcp_file,
                            family=args.ip_family,
                            dgn_dir=args.ip_dgn_dir,
                            jou_file=args.ip_dcp_jou_file,
                            log_file=args.ip_dcp_log_file)

        if not ip.dir:
            ip.dir = os.path.join(
                project.products_root, 'ip', '{vitis.version}')

        if not ip.dgn_dir:
            ip.dgn_dir = os.path.join(
                project.build_root, 'ip', 'dgn', '{vitis.version}')
        elif not os.path.isabs (ip.dgn_dir):
            ip.dgn_dir = os.path.join(project.build_root, ip.dgn_dir)

    if args.list:
        rc = 0
        if components is None or (len(components) == 0):
            components = ['*']

        workspace = Workspace.get(workspace)
        report(args.verbose, printer, project, workspace, components, 'List')
        cmp_paths = files.get_components(workspace, components)
        if len(cmp_paths):
            idx = 1
            for cmp_path in cmp_paths:
                cmp_name = os.path.split(cmp_path)[1]
                printer.item(idx, cmp_name, '')
                idx += 1
        else:
            noComponentsFound(printer, workspace, components)
            rc = -1
        printer.footer()

        return rc

    if options.stages == 0:
        print("ERROR: No action was specify\n"
              "       Specify one of: --csim, --synthesis, --cosim, "
              "--package, --implementation, --ip",
              file=sys.stderr)
        return -1

    run = Run(options, project.root, ip, printer)
    workspace = Workspace.get(workspace)

    report(args.verbose, printer, project, workspace, components, run)

    cmps = files.get_components(workspace, components)

    rc = 0
    if cmps:
        icmp = 1
        cmps = sorted(cmps)
        left = len(cmps)
        for cmp_path in cmps:
            printer.item(icmp, "Component", cmp_path)
            status = run.execute(cmp_path)
            if status:
                rc = status   # keep last non-zero so a failed stage propagates
            left -= 1
            icmp += 1
            if left and args.verbose:
                print()
    else:
        noComponentsFound(printer, workspace, components)

    printer.footer()
    return rc
# ------------------------------------------------------------------------------


if __name__ == '__main__':
    status = doit()
    sys.exit(status)
